beaker/ext/redisclusternm.py

```diff
@@ -0,0 +1,158 @@
+            raise RuntimeError('redis is not available')
+
+        if isinstance(url, string_type):
+            for url in url.split(','):
+                url_options = redis.connection.parse_url(url)
+                if 'db' in url_options:
+                    # Argument 'db' is not possible to use in cluster mode
+                    raise redis.cluster.RedisClusterException(
+                        "A ``db`` querystring option can only be 0 in cluster mode"
+                    )
+                self.nodes.append(redis.cluster.ClusterNode(
+                    host=url_options.get('host'),
+                    port=url_options.get('port')
+                ))
+            self.client = RedisClusterNamespaceManager.clients.get(
+                url, redis.RedisCluster, startup_nodes=self.nodes
+            )
+        else:
+            self.client = url
+
+    def _format_key(self, key):
+        if not isinstance(key, str):
+            key = key.decode('ascii')
+        if len(key) > (self.MAX_KEY_LENGTH - len(self.namespace) - len('beaker_cache:') - 1):
+            if not PY2:
+                key = key.encode('utf-8')
+            key = sha1(key).hexdigest()
+        return 'beaker_cache:%s:%s' % (self.namespace, key)
+
+    def get_creation_lock(self, key):
+        return RedisClusterSynchronizer(self._format_key(key), self.client)
+
+    def __getitem__(self, key):
+        entry = self.client.get(self._format_key(key))
+        if entry is None:
+            raise KeyError(key)
+        return pickle.loads(entry)
+
+    def __contains__(self, key):
+        return self.client.exists(self._format_key(key))
+
+    def has_key(self, key):
+        return key in self
+
+    def set_value(self, key, value, expiretime=None):
+        value = pickle.dumps(value)
+        if expiretime is None and self.timeout is not None:
+            expiretime = self.timeout
+        if expiretime is not None:
+            self.client.setex(self._format_key(key), int(expiretime), value)
+        else:
+            self.client.set(self._format_key(key), value)
+
+    def __setitem__(self, key, value):
+        self.set_value(key, value)
+
+    def __delitem__(self, key):
+        self.client.delete(self._format_key(key))
+
+    def do_remove(self):
+        for k in self.keys():
+            self.client.delete(k)
+
+    def keys(self):
+        return self.client.keys('beaker_cache:%s:*' % self.namespace)
+
+
+class RedisClusterSynchronizer(SynchronizerImpl):
+    """Synchronizer based on redis cluster.
+
+    Provided ``url`` can be both a redis connection string or
+    an already existing RedisCluster instance.
+
+    This Synchronizer only supports 1 reader or 1 writer at time, not concurrent readers.
+    """
+    # If a cache entry generation function can take a lot,
+    # but 15 minutes is more than a reasonable time.
+    LOCK_EXPIRATION = 900
+    MACHINE_ID = machine_identifier()
+
+    def __init__(self, identifier, url):
```

Comment:
is it one url or several urls? If can be a list of urls, then add similar comment like for  RedisClusterNamespaceManager

Reply:
It is several urls, applied same suggestion as above.